emukit/core/optimization/local_search_acquisition_optimizer.py

```diff
@@ -0,0 +1,110 @@
+import logging
+from typing import List, Tuple
+
+import numpy as np
+
+from emukit.core import (CategoricalParameter, OneHotEncoding, OrdinalEncoding,
+                         ParameterSpace)
+from emukit.core.acquisition import Acquisition
+from emukit.core.optimization.acquisition_optimizer import \
+    AcquisitionOptimizerBase
+from emukit.experimental_design.model_free.random_design import RandomDesign
+
+_log = logging.getLogger(__name__)
+
+
+class LocalSearchAcquisitionOptimizer(AcquisitionOptimizerBase):
+    """ Optimizes the acquisition function """
+    def __init__(self, space: ParameterSpace, num_steps: int, num_samples: int, **kwargs) -> None:
+        """
+        :param space: The parameter space spanning the search problem.
+        :param num_steps: Maximum number of steps to follow from each start point.
+        :param num_samples: Number of initial sampled points where the local search starts.
+        """
+        self.space = space
+        self.num_steps = num_steps
+        self.num_samples = num_samples
+        self.sampler = RandomDesign(space)
+
+    def _neighbours_per_parameter(self, all_features: np.ndarray) -> List[np.ndarray]:
+        assert all_features.ndim == 1, "Expected 1d array, got {}d".format(all_features.ndim)
+
+        parameters = self.space.parameters
+        neighbours = []
+        current_feature = 0
+        for parameter in parameters:
+            features = parameter.round(
+                all_features[current_feature:(current_feature + parameter.dimension)]
+                .reshape(1, -1)).ravel()
+            if isinstance(parameter, CategoricalParameter):
+                if isinstance(parameter.encoding, OrdinalEncoding):
```

Comment:
Shall we move this to a `get_neighbours` method of the encoding class? Also, should we be able to deal with discrete parameters in this bit too?

Reply:
Not too sure here. This method seems very specific, having many specific methods like that in Encoding or other general purpose interfaces may make it hard to provide custom implementations of these classes later

Reply:
Initially I prefered moving to the parameter. But I cannot think of many more usecases for one-exchange neighbourhoods. Therefore I would prefer keeping this in the optimizer for now and move them to a separate util class/module if more usecases occur.

Reply:
> Also, should we be able to deal with discrete parameters in this bit too?
Discrete ones could be handled like ordinal (select the one lower and higher).

We also should be able to handle continuous parameters: E.g sample k points from a normal distribution (mean at the current, std as hyperparameter e.g. 0.2).

Reply:
Discrete and continuous parameters are now also handled.